src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -416,25 +382,41 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
 
                     int chunkBytes = Math.Min(count, _sslState.MaxDataSize);
                     int encryptedBytes;
-                    SecurityStatusPal status = _sslState.EncryptData(buffer, offset, chunkBytes, ref outBuffer, out encryptedBytes);
-                    if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                    byte[] outBuffer = s_PinnableWriteBufferCache.AllocateBuffer();
+                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
                     {
-                        // Re-handshake status is not supported.
-                        ProtocolToken message = new ProtocolToken(null, status);
-                        throw new IOException(SR.net_io_encrypt, message.GetException());
+                        PinnableBufferCacheEventSource.Log.DebugMessage3("In System.Net._SslStream.StartWriting Using Pinnable", this.GetHashCode(), count, PinnableBufferCacheEventSource.AddressOfByteArray(outBuffer));
                     }
 
-                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                    try
+                    {
+                        SecurityStatusPal status = _sslState.EncryptData(buffer, offset, chunkBytes, ref outBuffer, out encryptedBytes);
+
+                        if (status.ErrorCode == SecurityStatusPalErrorCode.ContinueNeeded)
+                        {
+                            chunkBytes = 0;
+                        }
+                        else if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                        {
+                            // Re-handshake status is not supported.
+                            ProtocolToken message = new ProtocolToken(null, status);
+                            throw new IOException(SR.net_io_encrypt, message.GetException());
+                        }
+
+                    }
+                    catch
                     {
-                        PinnableBufferCacheEventSource.Log.DebugMessage3("In System.Net._SslStream.StartWriting Got Encrypted Buffer",
-                            this.GetHashCode(), encryptedBytes, PinnableBufferCacheEventSource.AddressOfByteArray(outBuffer));
+                        FreeBuffer(outBuffer);
+                        throw;
                     }
 
+
                     if (asyncRequest != null)
                     {
                         // Prepare for the next request.
                         asyncRequest.SetNextRequest(buffer, offset + chunkBytes, count - chunkBytes, s_resumeAsyncWriteCallback);
-                        Task t = _sslState.InnerStream.WriteAsync(outBuffer, 0, encryptedBytes);
+                        Task t = _sslState.InnerStream.WriteAsync(outBuffer, 0, encryptedBytes)
+                            .ContinueWith(_freeBufferAction, outBuffer);
```

Comment:
>  I wanted to either a. rip out the APM

As a subsequent change, I would be all for seeing the APM implementation overhauled / gutted, as long as we pay attention to perf in the process, which I know we will :)